### PR TITLE
[cffLib] save file position against lazy-load of charset

### PR DESCRIPTION
**Pull request by derwind**
This PR is related to #1177.

As is discussed, parseEncoding0 may receive unexpected file position due to lazy-load of 'charset'.
So I save file position before calling parseEncoding0 and restore the position at the beginning of parseEncoding0.

Similar modification is applied to parseCharset0 because of symmetry. Actually it may not be needed since 'strings' which has possibility of lazy-load seems to be always loaded before calling parseCharset0.

### CODE REVIEW

**Comment by anthrotype** [comment]
Thanks for the PR!

That's not exactly what I had in mind though.

The idea was to

> make such behind-the-scene lazy loads always seek the file back where it was...

So, in each of the `*Converter` classes's `read` methods, before they actually do any `file.seek` or `file.read` (either themselves or some other functions which they call), we would `tell` the file's current position, save to a local variable `pos`, then at the end of a converter's `read` method, seek back to the original saved position, so that the callers can call a converter's `read` method in any order and it's guaranteed to leave the file in the position that it found it.

You don't need to change the signatures of parseEncoding or parseCharset to add an extra `pos` argument. Simply do a file.seek(pos) at the end of the `*Converter.read` methods.

**Comment by derwind** [comment]
Thank you for your comments!

I understand and I will update codes again and test them. Indeed it is better if the signatures are unchanged.

**Comment by derwind** [comment]
I modified my codes again. I confirmed the result of reading TopDict.Encoding was as expected. Also I confirmed ttx command dumped same XMLs as those before creating PR by using SourceSansPro-Regular.otf and SourceHanSans-Regular.otf.

**Comment by anthrotype** [comment]
Thanks!

I have another idea (sorry if I keep you editing the code).

What if in the base class for all the Converters we define a `read` method and another one called `_read` which subclasses should implement (which does the actual reading). The `read` method of the superclass calls the _read method, but before and after ensures that the file position doesn't change.

```python
def read(self, ...):
    pos = file.tell()
    try:
        self._read(...)
    finally:
        file.seek(pos)

def _read(self, ...):
    raise NotImplementedError

```

**Comment by derwind** [comment]
@anthrotype Thank you for your advice again!
It is a little difficult for me to decide which is elegant but I implemented your advice.

On of the merits of previous implementation is that one can easily understand the intention when reading `*Converter.read`. But one should pay attention to possible lazy-loads when implementing new Converter.
On the other hand one of the merits of this implementation is that one can easily introduce new Converter without being conscious of lazy-load. But one may feel it difficult to notice the reason why `_read` methods were introduced when he or she sees them for the first time.

This seems a trade-off.

File: Lib/fontTools/cffLib/__init__.py [hunk 1]
```diff
@@ -1660,7 +1673,7 @@ def xmlRead(self, name, attrs, content, parent):
 
 class VarStoreConverter(SimpleConverter):
```
Comment (anthrotype) [inline Lib/fontTools/cffLib/__init__.py]: you forgot this one?

File: Lib/fontTools/cffLib/__init__.py [hunk 2]
```diff
@@ -1027,9 +1027,22 @@ def buildConverters(table):
 	return d
 
 
-class SimpleConverter(object):
+class BaseConverter(object):
```
Comment (anthrotype) [inline Lib/fontTools/cffLib/__init__.py]: SimpleCoverter is really the BaseConverter, as it's only used as base class and never used as such. so you can define read and _read on SimpleConverter itself.

**Comment by anthrotype** [comment]
another merit of this is that `finally` branch is guaranteed to be called no matter what happens in the `try` block, whereas in the previous one if an exception occurred the file would not rewind itself.

File: Lib/fontTools/cffLib/__init__.py (hunk 1 above)
Comment (derwind) [reply Lib/fontTools/cffLib/__init__.py]: According to your next comment I will update SimpleConverter itself.

File: Lib/fontTools/cffLib/__init__.py [hunk 3]
```diff
@@ -1027,20 +1027,16 @@ def buildConverters(table):
 	return d
 
 
-class BaseConverter(object):
+class SimpleConverter(object):
 
 	def read(self, parent, value):
+		if not hasattr(parent, "file"):
```
Comment (anthrotype) [inline Lib/fontTools/cffLib/__init__.py]: good catch

File: Lib/fontTools/cffLib/__init__.py [hunk 4]
```diff
@@ -1027,20 +1027,16 @@ def buildConverters(table):
...
+			return self._read(parent, value)
 		file = parent.file
 		pos = file.tell()
-		try:
```
Comment (anthrotype) [inline Lib/fontTools/cffLib/__init__.py]: why did you remove the try... finally... block?

Comment (derwind) [reply Lib/fontTools/cffLib/__init__.py]: Now SimpleConverter is the base class from which other Converters are derived, and the default meaningful _read method always exists and it simply returns 'value'. So I removed try finally block.

Comment (anthrotype) [reply Lib/fontTools/cffLib/__init__.py]: No, you need to put it back. The function of the finally is to ensure the statement after it will be called at the end whether the block inside the try completed successfully or it raised an exception.

Comment (derwind) [reply Lib/fontTools/cffLib/__init__.py]: Uh... try block should exist with considering overridden _read throw any exception ... sorry I will restore the try finally block.

Comment (anthrotype) [reply Lib/fontTools/cffLib/__init__.py]: yep. thanks

File: Lib/fontTools/cffLib/__init__.py (hunk 3 above)
Comment (derwind) [reply Lib/fontTools/cffLib/__init__.py]: Thanks!

**Comment by anthrotype** [comment]
one last optimization (but we can do it another time) could be to identify which of the Converter classes take a `parent` argument with a `file` attribute, and which ones don't; and only for the former do the wrapped _read; the latter don't manipulate files so they can implement `read` and be done with it (avoiding the second method call).
Because I imagine that this `parent` thing either always has a `file` or `never`, and that specific converters are always used with the same type of `parent` objects.
Anyway, thanks again. Merging now

**Comment by derwind** [comment]
@anthrotype Thank you for your patientlly reviewing my codes many times!

**Comment by anthrotype** [comment]
> I imagine that this parent thing either always has a file or never,

that's not true, actually. Both Index and BaseDict instances can be either of two state: they may or may not have a `file` attribute, depending on whether they have been initialized from a binary font file or whether they started empty and populated later. So the code is fine like this.

**Comment by behdad** [comment]
I would have guessed we put the seek logic in `__getattr__`. This works too though.

**Comment by anthrotype** [comment]
@behdad damn, you're right! Why didn't I think of that... :duck:

**Comment by derwind** [comment]
It's interesting! I implemented @behdad's idea as below and tested it. It worked fine!

```
$ git diff 4fec016862324092bf7d4643eb694903ffc114ac Lib/fontTools/cffLib/__init__.py
diff --git a/Lib/fontTools/cffLib/__init__.py b/Lib/fontTools/cffLib/__init__.py
index a11b8f6a..b7c4f584 100644
--- a/Lib/fontTools/cffLib/__init__.py
+++ b/Lib/fontTools/cffLib/__init__.py
```
```python
@@ -2234,7 +2234,10 @@ class BaseDict(object):
                if value is None:
                        raise AttributeError(name)
                conv = self.converters[name]
+               pos = self.file.tell() if hasattr(self, "file") else None
                value = conv.read(self, value)
+               if pos is not None:
+                       self.file.seek(pos)
                setattr(self, name, value)
                return value
```

If comparing `__getattr__` to 'a parent' and `*Converter.read` to 'his or her child', previous implementation seems as if the child who cluttered his or her toys put the toys away by his or herself, and newer implementation seems as if the child left his or her toys lying about and the parent put the toys away on behalf of the child.

**Comment by anthrotype** [comment]
Haha! Nice metaphor :)
If that is the only place where the read methods are called from, then it makes more sense to have it that way (i’d still wrap the call to read in a try finally block).
I’ll merge if you send a pull request, thanks

**Comment by derwind** [comment]
> If that is the only place where the read methods are called from

Surely if 'the spoiled child' is only one, 'the parent' can easily take care of 'the child'. But if there are many such children, 'the parent' may fail to take care.
And now I am worried about 'read' call in `FontDictCompiler.__init__` and I don't know when this path is executed and whether conv has 'file' or not.
```python
				default = dictObj.defaults.get(name)
				if default is not None:
					conv = dictObj.converters[name]
					default = conv.read(dictObj, default)
```

Considering this case, previously merged patch seems to me safer because all 'children' are always well-behaved :)
